src/platforms/ctxlink/ctxLink_traceswo.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Based on work that is Copyright (C) 2017 Black Sphere Technologies Ltd.
+ * Based on work that is Copyright (C) 2017 Dave Marples <dave@marples.net>
+ * Copyrigh (C) Sid Price 2020 <sid@sidprice.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.	 If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file implements capture of the TRACESWO output using ASYNC signalling.
+ *
+ * ARM DDI 0403D - ARMv7M Architecture Reference Manual
+ * ARM DDI 0337I - Cortex-M3 Technical Reference Manual
+ * ARM DDI 0314H - CoreSight Components Technical Reference Manual
+ */
+
+/* TDO/TRACESWO signal comes into the SWOUSART RX pin.
+ */
+
+#include "general.h"
+// #include "cdcacm.h"
+#include "traceswo.h"
+#include "platform.h"
+#include "WiFi_Server.h"
+
+#include <libopencm3/cm3/common.h>
+#include <libopencmsis/core_cm3.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/timer.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/usart.h>
+
+/* For speed this is set to the USB transfer size */
+#define FULL_SWO_PACKET (64)
+/* Default line rate....used as default for a request without baudrate */
+#define DEFAULTSPEED (2250000)
+
+#define BUFFER_SIZE 1024
+
+static volatile uint32_t input_buffer = 0;  // input buffer index
+static volatile uint32_t output_buffer = 0; // output buffer index
+volatile uint32_t buffer_size = 0;          // Number of bytes in the buffer
+
+static uint8_t trace_rx_buf[BUFFER_SIZE] = {0};
+#define NUM_PINGPONG_BUFFERS 2
+static uint8_t pingpong_buffers[NUM_PINGPONG_BUFFERS * FULL_SWO_PACKET] = {0};
+static uint32_t buffer_select = 0;
+//
+// Check for SWO Trace nework client, if present send
+// any queued data
+//
+static uint8_t swo_data[BUFFER_SIZE];
+
+void trace_send_data(void)
+{
+	if (is_swo_trace_client_connected()) {
+		uint32_t dataCount;
+		__atomic_load(&buffer_size, &dataCount, __ATOMIC_RELAXED);
+		if (dataCount >= FULL_SWO_PACKET) {
+			//
+			// Copy the data
+			//
+			for (uint32_t i = 0; i < dataCount; i++) {
+				swo_data[i] = trace_rx_buf[output_buffer++];
+				if (output_buffer >= BUFFER_SIZE) {
+					output_buffer = 0;
+				}
+			}
+			send_swo_trace_data(&swo_data[0], dataCount);
+			__atomic_fetch_sub(&buffer_size, dataCount, __ATOMIC_RELAXED);
+		}
+	}
+}
+
+void _trace_buf_drain(usbd_device *dev, uint8_t ep)
```

Comment:
Defining `_`-prefixed functions in user code invoke UB, please don't do that.

Reply:
This file is likely to have major changes when the Wi-Fi SWO feature is integrated with the new SWO structure in the base platform. This function is currently unused, therefore exckude from compilation, but left here for reference.

Reply:
That's noted 👍🏼 couldn't remember if this was being compiled in or not so reviewed as if it was.